File: ios/ExpoTailscaleCheckModule.swift
import ExpoModulesCore

public class ExpoTailscaleCheckModule: Module {
  private var detector: TailscaleDetector?
  private var state: TailscaleState?
  private let stateLock = NSLock()

  public func definition() -> ModuleDefinition {
    Name("ExpoTailscaleCheck")

    OnCreate {
      detector = TailscaleDetector()

      detector?.onStateChanged = { [weak self] state in
        guard let self = self else { return }

        self.stateLock.lock()
        let changed = self.state != state
        if changed { self.state = state }
        self.stateLock.unlock()

        if changed {
          self.sendEvent("onStateChange", [
            "state": state.rawValue
          ])
        }
      }
    }

    OnDestroy {
      detector = nil
    }

    Events("onStateChange")

    Function("getState") {
      self.stateLock.lock()
      let current = self.state
      self.stateLock.unlock()
      return current?.rawValue ?? "UNKNOWN"
    }
    
    Function("tailscaleInterface") { () -> [String: String]? in
      guard let interface = detector?.tailscaleInterface() else { return nil }
      return ["name": interface.name, "ip": interface.ip]
    }
  }
}

File: android/src/main/java/expo/modules/tailscalecheck/ExpoTailscaleCheckModule.kt
package expo.modules.tailscalecheck

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoTailscaleCheckModule : Module() {
  private var detector: TailscaleDetector? = null

  override fun definition() = ModuleDefinition {
    Name("ExpoTailscaleCheck")

    OnCreate {
      val context = appContext.reactContext ?: return@OnCreate

      detector = TailscaleDetector(context) { state ->
        sendEvent(
          "onStateChange",
          mapOf(
            "state" to getTailscaleStateValue(state)
          )
        )
      }
      detector?.start()
    }

    OnDestroy {
      detector?.stop()
      detector = null
    }

    Events("onStateChange")

    Function("getState") {
      val state = detector?.getCurrentState()
      getTailscaleStateValue(state)
    }

    Function("tailscaleInterface") {
      val tsInterface = detector?.tailscaleInterface()
      tsInterface
    }

    Function("openTailscaleApp") { packageName: String? ->
      val context = appContext.reactContext

      if (context != null) openTailscaleApp(context, packageName)
    }

    Function("connectVPN") { packageName: String? ->
      val context = appContext.reactContext

      if (context != null) connectVPN(context, packageName)
    }

    Function("disconnectVPN") { packageName: String? ->
      val context = appContext.reactContext

      if (context != null) disconnectVPN(context, packageName)
    }
  }
}
